Userland/Libraries/LibWeb/IndexedDB/IDBFactory.cpp

```diff
@@ -25,4 +33,64 @@ void IDBFactory::initialize(JS::Realm& realm)
     WEB_SET_PROTOTYPE_FOR_INTERFACE(IDBFactory);
 }
 
+// https://w3c.github.io/IndexedDB/#dom-idbfactory-open
+JS::ThrowCompletionOr<JS::NonnullGCPtr<IDBOpenDBRequest>> IDBFactory::open(String const& name, Optional<u64> version)
+{
+    auto& realm = this->realm();
+
+    // 1. If version is 0 (zero), throw a TypeError.
+    if (version.has_value() && version.value() == 0)
+        return realm.vm().throw_completion<JS::TypeError>(JS::ErrorType::InvalidVersion, 0);
+
+    // 2. Let environment be this's relevant settings object.
+    auto& environment = HTML::relevant_settings_object(*this);
+
+    // 3. Let storageKey be the result of running obtain a storage key given environment.
+    //    If failure is returned, then throw a "SecurityError" DOMException and abort these steps.
+    auto storage_key = StorageAPI::obtain_a_storage_key(environment);
+    if (!storage_key.has_value())
+        return realm.vm().throw_completion<WebIDL::SecurityError>("Failed to obtain a storage key"_string);
```

Comment:
I believe this should be a DOMException instead of a throw completion. In general, I believe Web specs should be returning an ExceptionOr instead of a JS::ThrowCompletion

Reply:
Agree on this. Makes it a bit fun when the spec says to "throw a TypeError" but there should be a few examples of how to shove one of those into a WebIDL::ExceptionOr.